numpy/core/src/multiarray/getset.c

```diff
@@ -513,6 +513,7 @@ array_descr_set(PyArrayObject *self, PyObject *arg, void *NPY_UNUSED(ignored))
         /* resize on last axis only */
         int axis = PyArray_NDIM(self) - 1;
         if (PyArray_DIMS(self)[axis] != 1 &&
+                PyArray_STRIDES(self)[axis] != 0 &&
```

Comment:
Maybe this should be
```suggestion
        if (PyArray_DIMS(self)[axis] > 1 &&
```

Reply:
This suggestion somewhat makes sense to me.  OTOH, I don't want anyone to set `arr.dtype` to begin with, and it would lead to non-zero strides.
So I think, I am fine with just forcing that one person who runs into it to doing weird stuff.  They should be doing different stuff anyway...

@eric-wieser I am planning to just merge thi sin a bit.  If you care and its merged, just make a note and I will follow-up.

Reply:
Since this makes sense let's put it in.

Reply:
Isn't `view` still implemented via this code (making a copy of the array container then setting `.dtype`?)

Reply:
Yeah, you are right `view` does indeed use this.  The tests are failing here for some reason though, I didn't investigate it yet:
```
../../builds/venv/lib/python3.8/site-packages/numpy-1.23.0.dev0+1181.gadd7834a6-py3.8-linux-x86_64.egg/numpy/linalg/tests/test_linalg.py:320: in _make_strided_cases
    for a, a_label in _stride_comb_iter(case.a):
        a          = array([[0.36561909],
       [0.8016026 ],
       [0.78273559],
       [0.70135538],
       [0.62277659]])
        a_label    = 'stride_xxx_0'
        b          = array([0.49368265, 0.8405377 , 0.71209699, 0.44390898, 0.03103486])
        b_label    = 'stride_-4'
        case       = <LinalgCase: 0x4>
        new_case   = <LinalgCase: 5x1_stride_xxx_0_stride_-4>
        new_cases  = [<LinalgCase: single_stride_+1_+1_stride_+1>, <LinalgCase: single_stride_+1_+1_stride_+3>, <LinalgCase: single_stride_...stride_+1_+3_stride_+1>, <LinalgCase: single_stride_+1_+3_stride_+3>, <LinalgCase: single_stride_+1_+3_stride_-4>, ...]
../../builds/venv/lib/python3.8/site-packages/numpy-1.23.0.dev0+1181.gadd7834a6-py3.8-linux-x86_64.egg/numpy/linalg/tests/test_linalg.py:291: in _stride_comb_iter
    xi.view(np.uint32).fill(0xdeadbeef)
E   ValueError: To change to a dtype of a different size, the last axis must be contiguous
        new_shape  = [0, 4]
        repeats    = (1, 1)
        slices     = (slice(None, None, 1), slice(None, None, 1))
        stride_set = [(1, 3, -4), (1, 3, -4)]
        x          = array([], shape=(0, 4), dtype=float64)
        xi         = array([], shape=(0, 4), dtype=float64)
```

Reply:
Reverted.

Reply:
I think it would be worth understanding the failure. The `PyArray_DIMS(self)[axis] != 1 && PyArray_STRIDES(self)[axis] != 0` logic feels wrong to me, as it means zero-size arrays may or may not reach this error depending on their stride (which can be anything as long as their allocation is ok).

Reply:
The reason is that we are inspecting the stride of a `(0, 4)` array.  We are setting the strides to `(0, 0)`, so the second dimension is "not contiguous" (not `stride == itemsize length <= 1`).
The fix might be to do a check for whether the array itself is contiguous (which captures the empty part).

Reply:
But, agruably, the current code is no better, since it would allow overriding a stride 0 dimension and giving it a non-zero stride, which could access out-of-bounds memory:
```
arr = np.ones(1, dtype="i8")
arr = np.broadcast_to(arr, 10)
arr.dtype = "i1"
print(arr)  # oops
```
(Somehow I had initially read it as being more strict, but it is the opposite.)